File: restim-maui/Core/Hdlc.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace RestimMaui.Core
{
    public class Hdlc
    {
        private const byte FRAME_BOUNDARY_MARKER = 0x7E;
        private const byte ESCAPE_MARKER = 0x7D;
        private const byte ESCAPE_XOR = 0x20;

        private bool _escapeNext;
        private readonly List<byte> _pendingPayload = new List<byte>();
        private readonly int _maxLen;

        public Hdlc(int maxLen = 1024)
        {
            _maxLen = maxLen;
        }

        public List<byte[]> Parse(byte[] data)
        {
            var resultingFrames = new List<byte[]>();

            foreach (var b in data)
            {
                if (b == FRAME_BOUNDARY_MARKER)
                {
                    if (_pendingPayload.Count >= 2)
                    {
                        var payload = _pendingPayload.GetRange(0, _pendingPayload.Count - 2).ToArray();
                        var computedCrc = Crc16(payload);
                        
                        // CRC is little-endian in the stream
                        var packetCrc = _pendingPayload[_pendingPayload.Count - 2] | (_pendingPayload[_pendingPayload.Count - 1] << 8);

                        if (computedCrc == packetCrc)
                        {
                            resultingFrames.Add(payload);
                        }
                    }
                    Reset();
                }
                else if (b == ESCAPE_MARKER)
                {
                    _escapeNext = true;
                }
                else
                {
                    var val = b;
                    if (_escapeNext)
                    {
                        val ^= ESCAPE_XOR;
                        _escapeNext = false;
                    }

                    _pendingPayload.Add(val);

                    if (_pendingPayload.Count > _maxLen)
                    {
                        Console.WriteLine("HDLC: max length exceeded, resetting buffer");
                        Reset();
                    }
                }
            }

            return resultingFrames;
        }

        public static byte[] Encode(byte[] payload)
        {
            if (payload.Length > 65536)
                throw new ArgumentException("Maximum length of payload is 65536");

            var checksum = Crc16(payload);
            var checksumBytes = new byte[] { (byte)(checksum & 0xFF), (byte)((checksum >> 8) & 0xFF) };

            var escapedPayload = Escape(payload);
            var escapedChecksum = Escape(checksumBytes);

            var output = new List<byte>();
            output.Add(FRAME_BOUNDARY_MARKER);
            output.AddRange(escapedPayload);
            output.AddRange(escapedChecksum);
            output.Add(FRAME_BOUNDARY_MARKER);

            return output.ToArray();
        }

        private static byte[] Escape(byte[] data)
        {
            var outList = new List<byte>();
            foreach (var b in data)
            {
                if (b == FRAME_BOUNDARY_MARKER || b == ESCAPE_MARKER)
                {
                    outList.Add(ESCAPE_MARKER);
                    outList.Add((byte)(b ^ ESCAPE_XOR));
                }
                else
                {
                    outList.Add(b);
                }
            }
            return outList.ToArray();
        }

        private void Reset()
        {
            _escapeNext = false;
            _pendingPayload.Clear();
        }

        // CRC-16-CCITT (X.25) polynomial 0x1021
        private static ushort Crc16(byte[] data)
        {
            ushort crc = 0xFFFF;
            foreach (var b in data)
            {
                var x = (ushort)((crc >> 8) ^ b);
                x ^= (ushort)(x >> 4);
                crc = (ushort)((crc << 8) ^ (x << 12) ^ (x << 5) ^ x);
            }
            return (ushort)(crc ^ 0xFFFF); 
            // Note: The python js-crc/models crc_16_x_25 implementation might differ slightly.
            // Standard X.25 often negates the output.
            // We should verify if this matches the TypeScript `crc_16_x_25`.
            // TS used `js-crc`.
            // If this implementation is incorrect, we can swap it. 
            // The standard CRC16-X25 logic is usually: Init FFFF, Poly 1021, RefIn True, RefOut True, XorOut FFFF.
            // The simple loop above is standard CCITT. 
            
            // Re-implementing essentially the same logic as typical X.25 to be safe:
            // return Crc16X25(data); 
        }
    }
}

File: foc-companion/lib/core/hdlc.dart
import 'dart:typed_data';
import 'package:foc_companion/services/app_logger.dart';

final _log = AppLogger.instance;

class Hdlc {
  static const int frameBoundaryMarker = 0x7E;
  static const int escapeMarker = 0x7D;
  static const int escapeXor = 0x20;

  bool _escapeNext = false;
  List<int> _pendingPayload = [];
  final int _maxLen;

  Hdlc({int maxLen = 1024}) : _maxLen = maxLen;

  List<Uint8List> parse(Uint8List data) {
    List<Uint8List> resultingFrames = [];

    for (int b in data) {
      if (b == frameBoundaryMarker) {
        if (_pendingPayload.length >= 2) {
          // Check CRC
          // CRC is little-endian in packet
          int packetCrc = _pendingPayload[_pendingPayload.length - 2] |
              (_pendingPayload[_pendingPayload.length - 1] << 8);

          Uint8List payload = Uint8List.fromList(
              _pendingPayload.sublist(0, _pendingPayload.length - 2));
          
          int computedCrc = _crc16(payload);

          if (computedCrc == packetCrc) {
            resultingFrames.add(payload);
          } else {
            _log.w("HDLC: CRC Mismatch! Computed: 0x${computedCrc.toRadixString(16)}, Packet: 0x${packetCrc.toRadixString(16)}");
          }
        }
        _reset();
      } else if (b == escapeMarker) {
        _escapeNext = true;
      } else {
        int val = b;
        if (_escapeNext) {
          val ^= escapeXor;
          _escapeNext = false;
        }

        _pendingPayload.add(val);

        if (_pendingPayload.length > _maxLen) {
          _log.w('HDLC: max length exceeded, resetting buffer');
          _reset();
        }
      }
    }

    return resultingFrames;
  }

  static Uint8List encode(Uint8List payload) {
    if (payload.length > 65536) {
      throw Exception("Maximum length of payload is 65536");
    }

    int checksum = _crc16(payload);
    Uint8List checksumBytes = Uint8List(2);
    checksumBytes[0] = checksum & 0xFF;
    checksumBytes[1] = (checksum >> 8) & 0xFF;

    List<int> escapedPayload = _escape(payload);
    List<int> escapedChecksum = _escape(checksumBytes);

    List<int> output = [];
    output.add(frameBoundaryMarker);
    output.addAll(escapedPayload);
    output.addAll(escapedChecksum);
    output.add(frameBoundaryMarker);

    return Uint8List.fromList(output);
  }

  static List<int> _escape(Uint8List data) {
    List<int> out = [];
    for (int b in data) {
      if (b == frameBoundaryMarker || b == escapeMarker) {
        out.add(escapeMarker);
        out.add(b ^ escapeXor);
      } else {
        out.add(b);
      }
    }
    return out;
  }

  void _reset() {
    _escapeNext = false;
    _pendingPayload.clear();
  }

  // CRC-16/X-25 (Poly 0x1021, Init 0xFFFF, RefIn/RefOut, XorOut 0xFFFF)
  // Implemented using reversed poly 0x8408 for LSB-first processing
  static int _crc16(Uint8List data) {
    int crc = 0xFFFF;
    for (int b in data) {
      crc ^= b;
      for (int i = 0; i < 8; i++) {
        if ((crc & 1) != 0) {
          crc = (crc >> 1) ^ 0x8408;
        } else {
          crc >>= 1;
        }
      }
    }
    return crc ^ 0xFFFF;
  }
}
